include/revng/Pipeline/Rank.h

```diff
@@ -16,4 +17,38 @@ class Rank : public DynamicHierarchy<Rank> {
   Rank(llvm::StringRef Name, Rank &Parent) : DynamicHierarchy(Name, Parent) {}
 };
 
+  template<StringTemplateParameter Str, typename LocationType, typename ParentRnk>
+  class RankDefinition : public Rank {
```

Comment:
1. What do you mean? I don't think `check-conventions` complained.

2. Yeah, I don't like it either BUT I'm 98%-ish sure that we want to keep the `Rank` as `Rank` since that name is used all over the pipeline. As for the `Definition`, we can choose whatever - these objects should always be constructed using a helper-function anyway.

Reply:
We usually go for `X` + `XBase` or `XImpl` + `X`.
If it's not supposed to be used explicitly by the user `XImpl` works.

Reply:
Anything that doesn't change the `Rank` name is fine by me

Reply:
1. Oh, wait, this seems to be the exact same issue we observed when seeing the namespaces being badly formatted after concepts.

Reply:
1 is done.
2 is still pending since I'm unsure about how we should name things 😓

Reply:
So what are the candidates as of now? Let's take a decision.

Reply:
Technically, there's no limit to how we might want to name them.

Seriously though, there are three different names we need:
1. The runtime rank facing the pipeline internals (I prefer this to stay as `Rank` after all, since this is the most "stable" interface)
2. The rank definition - inherits the `Rank` and extends it with a number of template parameters letting it be aware of its relations with other `Rank`s.
3. The `RankSpecialization` (name pending) concept - used in interfaces to make sure that a valid rank definition is passed to, for example `Location::at` function and the like. This is the most outwards facing interface a user is most likely to encounter.

What's your stance? Do you still prefer `RankImpl` for 2 or something?

Reply:
* `TypedRank` expresses the fact that it's not much more than a regular `Rank` but adds type safety, I like it.
* `RankDefinition` OK but it describes more what it is used for than what actually the thing *is*.
* `RankImpl` sad but OK

Pick one.

Reply:
`TypedRank` it is then.

Reply:
Done: https://github.com/revng/revng/pull/197/commits/bf0bab006f5a1fe707fc90ddc42f929b41dc61ee.